lib/Optimizer/TransformationPasses/EraseCopyOnStructExtractPass.cpp

```diff
@@ -0,0 +1,131 @@
+#include "GIL/InstVisitor.hpp"
+#include "GIL/Module.hpp"
+#include "GILGen/Context.hpp"
+#include "Instructions/Aggregates/StructExtractInst.hpp"
+#include "Instructions/Aggregates/StructFieldPtrInst.hpp"
+#include "Instructions/Memory/LoadInst.hpp"
+#include "Optimizer/AnalysisPasses/ValueUseChecker.hpp"
+#include "PassManager.hpp"
+
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace glu::optimizer {
+
+/// @class EraseCopyOnStructExtractPass
+/// @brief An optimization pass that transforms load [copy] + struct_extract
+/// patterns into struct_field_ptr + load [copy] patterns to avoid copying the
+/// entire struct.
+/// This pass transforms patterns like:
+///   %1 = load [copy] %0
+///   %2 = struct_extract %1
+/// into:
+///   %1 = struct_field_ptr %0
+///   %2 = load [copy] %1
+/// This avoids copying the entire struct when only one field is needed, while
+/// still properly copying the field if it has non-trivial ownership.
+class EraseCopyOnStructExtractPass
+    : public gil::InstVisitor<EraseCopyOnStructExtractPass> {
+private:
+    gil::Module *module;
+    std::optional<gilgen::Context> ctx = std::nullopt;
+
+    /// @brief Instructions to erase at the end
+    std::vector<gil::InstBase *> toErase;
+
+public:
+    /// @brief Constructor
+    EraseCopyOnStructExtractPass(gil::Module *module) : module(module) { }
+
+    /// @brief Visits a struct_extract instruction and tries to optimize the
+    /// pattern.
+    ///
+    /// @param extractInst The struct_extract instruction to visit
+    void visitStructExtractInst(gil::StructExtractInst *extractInst)
+    {
+        if (!ctx)
+            return;
+
+        // Get the struct value being extracted from
+        gil::Value structValue = extractInst->getStructValue();
+
+        // Check if the struct value comes from a load [copy] instruction
+        auto *definingInst = structValue.getDefiningInstruction();
+        if (!definingInst || !llvm::isa<gil::LoadInst>(definingInst)) {
+            return;
+        }
+
+        auto *loadInst = llvm::cast<gil::LoadInst>(definingInst);
+
+        // Only optimize load [copy] instructions
+        if (loadInst->getOwnershipKind() != gil::LoadOwnershipKind::Copy) {
+            return;
+        }
```

Comment:
why not optimize trivial ones as well?
you can't optimize [take] ones that's for sure, but trivial ones should be optimized, to avoid loading a huge value when it's not needed (although LLVM will certainly do it for you anyway)